# Обработка ошибок

## Try/catch

```js
async function go() {
  try {
    const pizza = await makePizza(["pepperoni"]);
    console.log(pizza);
  } catch (error) {
    console.log("Oh no!");
    console.error(error);
  }
}

go();
```

## Метод промисов catch

Мы всё равно используем async/await для получения результата, но обработчик ошибки помещаем в конец цепочки вызовов промисов в метод `catch()`

Обработчик ошибок внутри асинхронной функции

```js
function handleError(error) {
  console.log("Oh no!");
  console.error(error);
}

async function go() {
  const pizza = await makePizza(["pepperoni"]).catch(handleError);
  console.log(pizza);
}

go();
```

Обработчик ошибок на вызове асинхронной функции Так как `async` функция возвращает Promise

```js
function handleError(error) {
  console.log("Oh no!");
  console.error(error);
}

async function go() {
  window.doesNotExists();
  const pizza = await makePizza(["pepperoni"]);
  console.log(pizza);
}

go().catch(handleError);
```

### Функции высшего порядка

Почему так лучше? В случаях когда нужно запускать нашу функцию множество раз, может быть утомительно каждый раз дописывать ей обработку ошибок. Функция высшего порядка решает эту задачу для нас за один раз.

```js
function handleError(error) {
  console.log("Oh no!");
  console.error(error);
}

async function go() {
  window.doesNotExists();
  const pizza = await makePizza(["pepperoni"]);
  console.log(pizza);
}

// go().catch(handleError);

// Higher Order Function

function makeSafe(fn, handleError) {
  return function() {
    fn.catch(handleError);
  };
}

// go(); // unsafe будут ошибки

const safeGo = makeSafe(go, errorHandler);

safeGo();
```

## Async Prompts project

`ask()` — принимает объект опций и опциональное свойство cancel.

## Async Typer project


## Github Finder

[Регистрация приложения](https://github.com/settings/applications/new) для получения ключей

```js
class Github {
  constructor() {
    this.client_id = '',
    this.client_secret = ''
  }

  async getUser(user) {
    const profileResponse = await fetch(`https://api.github.com/users${user}?client_id=${this.client_id}&client_secret=${this.client_secret}`)
    const profile = await profileResponse.json();
    return {
      profile
    }
  }
}
```
